util/bloom/src/lib.rs

```diff
@@ -178,11 +174,11 @@ impl Bloom {
 		cmp::max(k_num, 1)
 	}
 
-	fn bloom_hash<T>(&self, hashes: &mut [u64; NUMBER_OF_HASHERS], item: &T, k_i: u32) -> u64
+	fn bloom_hash<T>(&self, hashes: &mut [u64; 2], item: &T, k_i: u32) -> u64
 		where T: Hash
 	{
-		if k_i < NUMBER_OF_HASHERS as u32 {
-			let mut sip = self.sips[k_i as usize].clone();
+		if k_i < 2 {
+			let mut sip = self.sip.clone();
 			item.hash(&mut sip);
 			let hash = sip.finish();
 			hashes[k_i as usize] = hash;
```

Comment:
The thing is that the `hashes` array will contain exactly the same number twice, so the whole thing is redundant as well.

I think we should initialize two different `SipHashers` and keep the logic here, but that will break backward compatibility.
If we want to use only a single hasher then this code should be simplified as well, and I think we should review if that hashing method doesn't generate too many collisions.

Reply:
Hi, could you explain in which way using different hashers will break backward compatibility, thank you.

Without the bc problem, using two different independent hashers is indeed the best choice according to Kirsch and Mitzenmacher's paper.

Reply:
A bloom filter of non-empty accounts is stored as part of the state and loaded via `Bloom::from_parts`. If we change the `bloom_hash` method it will affect `check` and therefore the restored `bitmap` will have no correlation to accounts we are checking it against.
The change is possible but will require a database migration similar to: https://github.com/paritytech/parity/blob/2b81b982197d27efeea80c8a176c385114d95fdb/ethcore/src/migrations/v10.rs#L17

Reply:
After speaking with some team members we decided that it would be best to leave the hashing logic as is (i.e. avoid migration) but simplify it as much as we can (perhaps gaining some performance).
So you can probably avoid calling `item.hash(&mut sip)` twice (just copy the value for the next iteration) and avoid storing the `SipHasher` alltogether (as it now suggests that the hasher state is actually maintained, which is not true because it's being cloned before use).

Reply:
@tomusdrw Thanks for explaining it.

I just updated the PR according to your suggestion.